src/bitprim/keoken/interpreter.cpp

```diff
@@ -0,0 +1,473 @@
+        istream_reader source(ds);
+
+        return version_dispatcher(block_height, tx, source);
+    }
+    return error::not_keoken_tx;
+}
+
+error_code_t interpreter::version_dispatcher(size_t block_height, transaction const& tx, reader& source) {
+
+    auto version = source.read_2_bytes_big_endian();
+    if ( ! source) return error::invalid_version_number;
+
+    switch (static_cast<version_t>(version)) {
+        case version_t::zero:
+            return version_0_type_dispatcher(block_height, tx, source);
+    }
+    return error::not_recognized_version_number;
+}
+
+error_code_t interpreter::version_0_type_dispatcher(size_t block_height, transaction const& tx, reader& source) {
+    auto type = source.read_2_bytes_big_endian();
+    if ( ! source) return error::invalid_type;
+
+    switch (static_cast<message_type_t>(type)) {
+        case message_type_t::create_asset:
+            return process_create_asset_version_0(block_height, tx, source);
+        case message_type_t::send_tokens:
+            return process_send_tokens_version_0(block_height, tx, source);
+    }
+    return error::not_recognized_type;
+}
+
+payment_address interpreter::get_first_input_addr(transaction const& tx) const {
+    auto const& owner_input = tx.inputs()[0];
+
+    output out_output;
+    size_t out_height;
+    uint32_t out_median_time_past;
+    bool out_coinbase;
+
+    if ( ! fast_chain_.get_output(out_output, out_height, out_median_time_past, out_coinbase, 
+                                  owner_input.previous_output(), libbitcoin::max_size_t, true)) {
+        return payment_address{};
+    }
+
+    return out_output.address();
+}
+
+error_code_t interpreter::process_create_asset_version_0(size_t block_height, transaction const& tx, reader& source) {
+    auto msg = message::create_asset::factory_from_data(source);
+    if ( ! source) return error::invalid_create_asset_message;
+
+    if (msg.amount() <= 0) {
+        return error::invalid_asset_amount;
+    }
+
+    auto owner = get_first_input_addr(tx);
+    if ( ! owner) {
+        return error::invalid_asset_creator;
+    }
+
+    state_.create_asset(msg.name(), msg.amount(), std::move(owner), block_height, tx.hash());
+    return error::success;
+}
+
+std::pair<payment_address, payment_address> interpreter::get_send_tokens_addrs(transaction const& tx) const {
+    auto source = get_first_input_addr(tx);
+    if ( ! source) {
+        return {payment_address{}, payment_address{}};
+    }
+
+    auto it = std::find_if(tx.outputs().begin(), tx.outputs().end(), [&source](output const& o) {
+        return o.address() && o.address() != source;
+    });
+
+    if (it == tx.outputs().end()) {
+        return {std::move(source), payment_address{}};        
+    }
+
+    return {std::move(source), it->address()};
+}
+
+error_code_t interpreter::process_send_tokens_version_0(size_t block_height, transaction const& tx, reader& source) {
+    auto msg = message::send_tokens::factory_from_data(source);
```

Comment:
missing msg construction validation

Reply:
fixed in `bitprim-core`
https://github.com/bitprim/bitprim-core/pull/37